core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -553,15 +558,20 @@ public static ReadBuilder read(InputFile file) {
   public static class ReadBuilder {
     private final InputFile file;
     private final Map<String, String> renames = Maps.newLinkedHashMap();
-    private ClassLoader loader = Thread.currentThread().getContextClassLoader();
+    private ClassLoader loader = null;
     private NameMapping nameMapping;
     private boolean reuseContainers = false;
     private org.apache.iceberg.Schema schema = null;
     private Function<Schema, DatumReader<?>> createReaderFunc = null;
     private BiFunction<org.apache.iceberg.Schema, Schema, DatumReader<?>> createReaderBiFunc = null;
     private final Function<Schema, DatumReader<?>> defaultCreateReaderFunc = readSchema -> {
       GenericAvroReader<?> reader = new GenericAvroReader<>(readSchema);
-      reader.setClassLoader(loader);
+      ClassLoader classLoader = loader;
+      if (classLoader == null) {
+        LOG.warn("Classloader for avro reader not set, use the thread context classloader.");
+        classLoader = Thread.currentThread().getContextClassLoader();
+      }
+      reader.setClassLoader(classLoader);
```

Comment:
@rdblue I'm not sure if we should just requireNonNull(loader) here.

Reply:
I don't think that we need to change any behavior in this class. The current behavior is adequate and setting the classloader is only needed if you are attempting to load a custom class or if the context classloader throws an unexpected exception, as in the Flink case.

Reply:
Adding warning mainly to prevent forgetting to set classloader if required. I've to permit that it is not perfect, but have no better idea.
It's ok to not change the behavior of this class.

I'll remove changes in this file.